src/Agent.Worker/Release/Artifacts/CustomArtifact.cs

```diff
@@ -0,0 +1,206 @@
+
+            if (customEndpoint == null)
+            {
+                throw new InvalidOperationException(StringUtil.Loc("RMCustomEndpointNotFound", agentArtifactDefinition.Name));
+            }
+
+            var details = JToken.Parse(customArtifactDetails).ToObject<CustomArtifactDetails>();
+            details.RelativePath = relativePath;
+            details.Endpoint = new ServiceEndpoint
+            {
+                Url = customEndpoint.Url,
+                Authorization = customEndpoint.Authorization
+            };
+            return details;
+        }
+
+
+        private async Task DownloadArtifact(
+            IExecutionContext executionContext,
+            IHostContext hostContext,
+            string localFolderPath,
+            CustomArtifactDetails customArtifactDetails,
+            CustomArtifactDownloadDetails artifact)
+        {
+            IDictionary<string, string> artifactTypeStreamMapping = customArtifactDetails.ArtifactTypeStreamMapping;
+            string streamType = GetArtifactStreamType(artifact, artifactTypeStreamMapping);
+
+            if (string.Equals(streamType, WellKnownStreamTypes.FileShare, StringComparison.OrdinalIgnoreCase))
+            {
+                var fileShareArtifact = new FileShareArtifact();
+                customArtifactDetails.RelativePath = artifact.RelativePath ?? string.Empty;
+                var location = artifact.FileShareLocation ?? artifact.DownloadUrl;
+                await fileShareArtifact.DownloadArtifactAsync(executionContext, hostContext, new ArtifactDefinition { Details = customArtifactDetails }, new Uri(location).LocalPath, localFolderPath);
+            }
+            else if (string.Equals(streamType, WellKnownStreamTypes.Zip, StringComparison.OrdinalIgnoreCase))
+            {
+                try
+                {
+                    IEndpointAuthorizer authorizer = SchemeBasedAuthorizerFactory.GetEndpointAuthorizer(
+                        customArtifactDetails.Endpoint,
+                        customArtifactDetails.AuthorizationHeaders);
+
+                    using (HttpWebResponse webResponse = GetWebResponse(executionContext, artifact.DownloadUrl, authorizer))
+                    {
+                        var zipStreamDownloader = HostContext.GetService<IZipStreamDownloader>();
+                        await zipStreamDownloader.DownloadFromStream(
+                            webResponse.GetResponseStream(),
+                            string.Empty,
+                            artifact.RelativePath ?? string.Empty,
+                            localFolderPath);
+                    }
+                }
+                catch (WebException)
+                {
+                    executionContext.Output(StringUtil.Loc("ArtifactDownloadFailed", artifact.DownloadUrl));
+                    throw;
```

Comment:
show we throw ArtifactDownloadException so that it doesn't get flagged as system errors in telemetry?